### DISCUSSION TITLE
[v4] How to handle css configuration for css variables defined based on a data-theme attribute

### DISCUSSION THREAD

**Discussion by benebene84** [tailwindlabs/tailwindcss#15199] [Help] [answered]
Hi everyone,

currently we are trying to migrate a design system to use the tailwind v4 beta. Up until now, the theming is done with css variable that are loaded based on a data-theme property on the body to enable different themes for this design system.

```html
<body data-theme="theme-xyz">
</body>
```

```css
[data-theme="theme-xyz"] {
...
--primary-main: #81ADC8;
...
}
```

In Tailwind v3 it was possible to define a custom color in the config file:
```ts
colors: {
'ds-primary-main': 'var(--primary-main)'
}
```

This also works if I use the @config directive and specify the old tailwind config. It generates the class and loads the variables based on the data-theme property.

However, if I try to migrate this to the css configuration, this is not working, because the css-variable is not known where it is referenced by tailwind:

<img width="251" alt="image" src="https://github.com/user-attachments/assets/0a2434f8-8862-4b9c-82dc-3531bf1798a8">

It works if I remove the data-theme selector and add them to `:root` in the css, but this would make it hard to maintain for applications relying on multiple themes.

Is there any way around it?

**Comment by adamwathan** [comment, marked as answer]
Hey! Use the `@theme inline` option (need to document this still), should help:

https://play.tailwindcss.com/1bFjrBuJMc?file=css

**Reply by benebene84** [reply]
Thanks for the swift reply! This works.

**Reply by datner** [reply]
with the `@theme` directive I half expected to have some building theme support like this
```css
@theme {
  --foo: 1;
}

@theme bar {
  --foo: 2;
}

/* or maybe */
@media theme(bar) {
  @theme {
    --foo: 2;
  }
}
```
to later do `theme-bar` or in SSR env work on imports. So like in page `/foo` only `tailwind.css` is imported with the `@theme` rule for  `--foo: 2;`, but `/bar` _also_ imports `bar.css` later, which defines a theme with `--foo: 1;`, and due to cascading it resolves or something

**Reply by datner** [reply]
Or with the existing directives (snooped into the codebase) then use `reference` somehow
